scripts/entity/components/movement_components/player_movement_component.gd

```diff
@@ -4,58 +4,96 @@ class_name PlayerMovementComponent extends BaseMovementComponent
 var _horizontal_deceleration: float = MAX_SPEED / DECELERATION_TIME
 
-# Called when the node enters the scene tree for the first time.
+## Moth Jump state variables
+var _is_moth_jumping: bool = false
+var _is_fluttering: bool = false
+var _moth_jump_timer: float = 0.0
+var _moth_jump_flutter_timer: float = 0.0
+var _moth_jump_initial_height: float = 0.0
+var _is_jump_button_held: bool = false
+
+
 func _ready() -> void:
-	#do nothing
 	pass
 
-
 ## Applies movement to the character.
 func _physics_process(delta: float) -> void:
 	super._physics_process(delta)
-	apply_gravity(delta)	
+	apply_gravity(delta)
 	apply_horizontal_movement(delta)
 	move_and_slide()
 
-
 func on_movement_input(horizontal_vector2: float) -> void:
 	_current_input_direction = horizontal_vector2
 
-
 func on_jump_input_started() -> void:
 	if is_on_floor():
 		velocity.y = JUMP_VELOCITY
 		on_jump_started.emit()
-
+	elif not _is_moth_jumping:
+		# Start Moth Jump
+		_is_moth_jumping = true
+		_is_fluttering = true
+		_moth_jump_timer = 0.0
+		_moth_jump_flutter_timer = 0.0
+		_moth_jump_initial_height = global_position.y
+		_is_jump_button_held = true
 
 func on_jump_input() -> void:
-	# do nothing
-	pass
-
+	_is_jump_button_held = true
 
 func on_jump_input_cancelled() -> void:
-	# do nothing
-	pass
-
+	_is_jump_button_held = false
 
 ## Moves the entity based on the current input direction.
 func apply_horizontal_movement(delta: float) -> void:
- 	# Direction button active
-	if _current_input_direction:
+	if _current_input_direction != 0:
 		# Accelerate towards the target speed
 		velocity.x = move_toward(velocity.x, _current_input_direction * MAX_SPEED, ACCELERATION * delta)
-	 # Direction button inactive
 	else:
 		# Decelerate towards zero speed
 		velocity.x = move_toward(velocity.x, 0, _horizontal_deceleration * delta)
+
+func apply_gravity(delta: float) -> void:
+	if _is_moth_jumping:
+		if _is_fluttering:
+			# Handle the flutter dip
+			var flutter_progress: float = _moth_jump_flutter_timer / MOTH_JUMP_FLUTTER_DURATION
+			# Copy in to desmos to see why I chose this function: -\sin\left(\frac{\left(x-a\right)^{2}}{c}\right)+1
+			var flutter_dip: float = -MOTH_JUMP_DIP_AMOUNT * (-sin((pow((flutter_progress-0.5),2))/0.16) + 1)
+			var desired_position_y: float = _moth_jump_initial_height + flutter_dip
+			var position_difference_y: float = desired_position_y - global_position.y
+			velocity.y = position_difference_y / delta
+			_moth_jump_flutter_timer += delta
+			if _moth_jump_flutter_timer >= MOTH_JUMP_FLUTTER_DURATION:
+				_is_fluttering = false
+				global_position.y = _moth_jump_initial_height
+				velocity.y = 0.0
+		else:
+			if _is_jump_button_held and _moth_jump_timer < MOTH_JUMP_HANGTIME:
+				# Maintain the current height
+				velocity.y = 0.0
+				_moth_jump_timer += delta
+			else:
+				# End Moth Jump
+				_is_moth_jumping = false
+				velocity.y += super.get_gravity().y * delta  # Resume normal gravity
```

Comment:
Id probably put this into a function or two and explain what you are doing. There is a lot of math but I cant tell you what goes to what easily.

Reply:
Refactored